### PR TITLE
Generalize sequence labeler and allow re-use embeddings for labeling

### PR DESCRIPTION
**Pull request by jlibovicky**
* sequence labeler can now use multiple inputs
* fixed bug with max length #790 
* labeler can reuse word embeddings for classification

### CODE REVIEW

File: neuralmonkey/decoders/sequence_labeler.py [hunk 1]
```diff
@@ -52,64 +62,36 @@ def target_tokens(self) -> tf.Tensor:
...
-        dweights_4d = tf.expand_dims(
-            tf.expand_dims(self.decoding_residual_w, 0), 0)
-
-        dmultiplication = tf.nn.conv2d(
-            embedded_inputs, dweights_4d, [1, 1, 1, 1], "SAME")
-        dmultiplication_3d = tf.squeeze(dmultiplication, axis=[2])
-
-        logits = multiplication_3d + dmultiplication_3d + biases_3d
-        return logits
+        return tf.layers.dense(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: Dropout na self.states?

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: jo vidim že tady je to teda opačně než by mělo - states by se měly asi vracet nedropovaný a tohle by si je mělo dropovat.

File: neuralmonkey/decoders/sequence_labeler.py [hunk 2]
```diff
@@ -142,6 +127,66 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+            embeddings = tf.stop_gradient(embeddings)
+
+        states = self.states
+        # pylint: disable=no-member
+        states_dim = self.states.get_shape()[-1].value
+        # pylint: enable=no-member
+        embedding_dim = self.embedded_sequence.embedding_sizes[0]
+        # pylint: disable=redefined-variable-type
+        if states_dim != embedding_dim:
+            states = tf.layers.dense(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: dropout na states?

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: dropout na embeddings?

File: neuralmonkey/runners/label_runner.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def __init__(self,
     def fetches(self) -> Dict[str, tf.Tensor]:
         return {
             "label_logprobs": self.decoder.logprobs,
-            "input_mask": self.decoder.encoder.input_sequence.temporal_mask,
+            "input_mask": self.decoder.encoders[0].temporal_mask,
```
Comment (jindrahelcl) [inline neuralmonkey/runners/label_runner.py]: možná bych dekodéru  (labeleru) dal tenzor input_mask, a ten bude dělat tadyto a k tomu ještě kontrlovat že ty masky jedno jsou.

File: neuralmonkey/decoders/sequence_labeler.py [hunk 4]
```diff
@@ -142,6 +127,66 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+                states, embedding_dim, name="project_for_embeddings")
+        # pylint: enable=redefined-variable-type
+
+        reshaped_states = tf.reshape(states, [-1, embedding_dim])
+        reshaped_logits = tf.matmul(
+            reshaped_states, embeddings, transpose_b=True, name="logits")
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Biasy necheme?

File: neuralmonkey/decoders/sequence_labeler.py [hunk 5]
```diff
@@ -142,6 +127,66 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+
+        self.embedded_sequence = embedded_sequence
+        self.train_embeddings = train_embeddings
+    # pylint: enable=too-many-arguments,too-many-locals
+
+    @tensor
+    def logits(self) -> tf.Tensor:
+        embeddings = self.embedded_sequence.embedding_matrix
+        if not self.train_embeddings:
+            embeddings = tf.stop_gradient(embeddings)
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Fakt muzes tohle udelat? Vzhledem k tomu, ze se jedna o fakticky posledni vrstvu, nestane se to, ze ti pri backpropu neprotece zadna informace do zbytku site (a tedy se nic nenaucis)?

File: neuralmonkey/decoders/sequence_labeler.py [hunk 6]
```diff
@@ -52,64 +62,36 @@ def target_tokens(self) -> tf.Tensor:
...
-            name="state_to_word_W",
-            shape=[self.rnn_size, len(self.vocabulary)])
-
     @tensor
-    def decoding_b(self) -> tf.Variable:
-        return get_variable(
-            name="state_to_word_b",
-            shape=[len(self.vocabulary)],
-            initializer=tf.zeros_initializer())
+    def concatenated_inputs(self) -> tf.Tensor:
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Kdyz budu mit vice enkoderu, co pracuji nad ruzne dlouhymi sekvencemi, tak to na tom concatu spadne, ne?
Nemela by se takova situace resit spise pres FactoredSequence/FactoredEncoder?

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: to je pravda

File: neuralmonkey/decoders/sequence_labeler.py [hunk 7]
```diff
@@ -52,64 +62,36 @@ def target_tokens(self) -> tf.Tensor:
...
     @tensor
-    def decoding_residual_w(self) -> tf.Variable:
-        input_dim = self.encoder.input_sequence.dimension
-        return get_variable(
-            name="emb_to_word_W",
-            shape=[input_dim, len(self.vocabulary)])
+    def states(self) -> tf.Tensor:
+        states = dropout(
+            self.concatenated_inputs, self.dropout_keep_prob, self.train_mode)
+
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: bude, ale v kombinaci třeba s RecurrentEncoder by tu pak zas nebyl ani jeden dropout. Je to teď v nekonzistentnim stavu, otázka je, jak to vyřešit. Já bych asi zatim předělal transformer, protože to bude míň práce a zůstane to v souladu s dosavadní ideou, že výstupy jsou pure numbers. Ale příliš si za tím nestojím, líbí se mi i idea, že každá komponenta ví sama nejlíp, za co má ten dropout dát.

[the remark this answers is not in the document]
Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: 1. Pouziva vubec nekdo recurrent encoder/decoder?
2. Jak budu popisovat dropout, trenovani apod., pokud budu mit >= 2 komponenty, ktere budou/nebudou dropoutovat stejnou vrstvu. Je to otazka predevsim multitask learningu (i kdyz pri nejakych rafinovanych kombinacich vice jak jednoho attention mechanismu tohle taky muze nastat). Ale pokud v budoucnu neplanujete podporovat multi-task learning, tak to asi muze byt jedno.
Bottom-line: stary approach k dropoutu je motivovany jenom jakymsi "estetickym citenim" (nejak jsem zatim nepostrehl zadny fakticky argument pro; ale mozna mi jen unikl). Argumenty pro dropout vystupu metody jsem popsal, viz vise.

[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: 1. jo, teď zrovna my
2. už jsi mě přesvědčil, viz #800

File: neuralmonkey/decoders/sequence_labeler.py (hunk 4 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: neděláme je ani jinde při `tie_embeddings`, viz autoregressive.py:225

File: neuralmonkey/decoders/sequence_labeler.py (hunk 5 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: neproteče tam přes tuhle lokální proměnnou `embeddings`, ale proteče tam skrz `states`.

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Jasne, uz to vidim

File: neuralmonkey/decoders/sequence_labeler.py (hunk 4 above)
Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Jo, nevsiml jsem si, ze EmbeddingsLabeler vzdycky vaze embeddingy.

**Comment by jindrahelcl** [comment]
Poznámku o bertovi jsem přehlídnul, s bertem jsme to nepoužívali... Máme normální language model, na kterej tenhleten labeler lepíme a je to pak skoro to samý, jako rekurentní dekodér.

**Comment by jindrahelcl** [comment]
Tak snad opraveno..

File: neuralmonkey/decoders/sequence_labeler.py [hunk 8]
```diff
@@ -120,14 +118,19 @@ def decoded(self) -> tf.Tensor:
...
-    def cost(self) -> tf.Tensor:
+    def train_xents(self) -> tf.Tensor:
         loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
             labels=self.train_targets, logits=self.logits)
 
         # loss is now of shape [batch, time]. Need to mask it now by
         # element-wise multiplication with weights placeholder
-        weighted_loss = loss * self.train_mask
-        return tf.reduce_sum(weighted_loss)
+        return loss * self.train_mask
```
Comment (varisd) [inline neuralmonkey/decoders/sequence_labeler.py]: Jo, jeste se mi tady nezda format train_xents.
Tady to vraci shape(batch, time), ale v autoregressivnich dekoderech to vraci shape(batch).

Bylo by teda fajn se dohodnout, co se bude vracet (bud tady vracet prumer-per-sequence; nebo v autoregressive by melo stacit vypnout average_across_timesteps v self.train_xents)

In the long run by samozrejme mel byt jeden spolecny predek "dekoder" s abstraktnima metodama, jako loss, xents apod., ale to bych klidne nechal do jineho PR.

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: V mariánovi to jde nastavovat ("ce-mean" vs "ce-mean-words"), s tím, že v nových modelech se používá loss zprůměrovanej ze všech slov v batchi, ale default má průměr po větách.

U labeleru dává víc smysl mít ten loss pro každej label zvlášť, kdežto u dekodéru asi spíš po větách, ale souhlasim, že se to má sjednotit. Je tim pádem asi lepší vracet (batch, time) kterej si pak zprůměruješ jak chceš.

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Taky jsem za druhou moznost, to ale znamena tedka jeste vypnout switch v AutoregressiveDecoder. Nevim, jak to ovlivni zabehle trenovani (predpokladam, ze minimalne; ale nemeril jsem to). Pokud udelas porovnani, tak to klidne muzes zamergovat timhle zpusobem.

Na druhou stranu uz mam vyzkousene ze udelat prumer pres vety a pak pres batch v seq. labeleru funguje, takze bych v tomto PR radej udelal tohle (hodil issue na poradne doreseni).

Kazdopadne by bylo fajn to uz ted mit v masteru sjednocene, nez to nekam zapadne. Klicove je, ze to vyrazne snizi uroven odrbavani v jinych komponentach, kde pak musis mit divne workaroundy tipu kontroly shapu, supported decoder apod.

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: autoregressive dekodér má nějakej switch? Vidim jen, že to dělá `reduce_mean` na `{train,runtime}_xents` (místo aby dělal mean jen přes validní pozice, ale jinak je to stejný)

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Mam na mysli toto:
https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/decoders/autoregressive.py#L291

seq2seq.sequence_loss ma prepinac average_across_timesteps, ktery je tedka True. Kdyby se vypnul, tak by to odpovidalo formatu train_xents v labeleru

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: už to vidim...

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: hotovo. musel se ještě změnit perplexity runner, kterej počítá s průměrama přes čas.
